### PR TITLE
NEW: Auto-injection into template

### PR DESCRIPTION
**Pull request by unclecheese**
This pull request removes the need to add `$BetterNavigator` to your template, and injects the html directly into the HTTPResponse after the template is parsed.

This is somewhat API-breaking. The old `BetterNavigator()` method still returns `DBHTMLText`, but just empty, so ensure that users who upgrade don't get two navigators.

Have also upgraded the framework dependency to 4.2, as 4.0 and 4.1 are EOL.

### CODE REVIEW

**Comment by jonom** [comment]
Thanks @unclecheese that's a sweet idea and I learned a neat trick today - didn't know about those extension points 😄. I didn't know BetterNavigator was used in the demo project either, so that's cool too.

I will make this a major version bump so would you mind removing the `BetterNavigator` method?

If a controller is sending an ajax response will this still work? Glancing at the regex I'm guessing that if there's no </body> tag then the better navigator output won't get injected so it should be okay? And I guess this shouldn't affect submitting a form either but did you test that?

The only other potential issue I can think of is that when `$BetterNavigator` was added to templates by the user, they could wrap that in an `<% if $Condition %>` to hide it. So maybe we need a `BetterNavigator.Enabled` config setting that can be modified in user code with `Config::inst()->set()` prior to the `afterCallActionHandler` hook being called to allow devs to opt out?

Lastly a bit of philosophising 🤔 I'm not sure how I feel about bumping the min framework requirement due to the EOL status. If this module is otherwise compatible with 4.0 I don't really see a reason to do that. This is a utility module so I don't feel it has a responsibility to enforce good security policies. Projects sometimes get stuck on EOL versions for all kinds of dumb reasons and I don't want to restrict access to this module because there's no budget for an upgrade or 4.2 broke some key function (I know, SemVer, but it happens...). Then again, if I'm doing a major version bump then it's a good time to make that change, so I guess I'm okay with it. </ 🤔>

**Comment by unclecheese** [comment]
Cheers, @jonom .. Good points. I'll address each of those:

* Have tested it with AJAX requests, and it does not inject due to the regex, as you saw. Could potentially add an `isAjax()` jut to be sure? It's not exactly unheard of to return a full body in an XHR, albeit rare.

* Conditional navigator -- fair point. I've added a `shouldShowNavigator(): bool` method that you can override in your controller to add custom logic. By default, it returns true.

* Major release: Technically the only API change is the type declarations. We could remove those and do this as a minor release. It's perfectly fine to bump the framework and PHP version for a minor release. We've done it in SS 4.4 because of EOLs. 

I'll leave you to make that call, though. Removing the type declarations on old methods, bumping the dependency constraints, and releasing as a minor feels like the best path from my view. Otherwise, as-is, major release.

File: composer.json [hunk 1]
```diff
@@ -16,7 +16,7 @@
         }
     ],
     "require": {
-        "silverstripe/framework": "^4.0"
+        "silverstripe/framework": ">= 4.2"
```
Comment (jonom) [inline composer.json]: Not sure how I feel about this. Wouldn't this promise compatibility with SS 5/6/7 etc.? A limitation of Composer (If I understand correctly) is that if I tag a release with this in it, I can't ever go back and update that tag if it turns out the code at that point is only compatible with SS4. I could tag a new release with a tighter version constraint, but previous release(s) would still exist and be seen as a viable options for SS5+ by Composer. I think ^4.2 would be safer?

File: src/Extension/BetterNavigatorExtension.php [hunk 2]
```diff
@@ -8,25 +8,44 @@
...
+    /**
+     * @var string
+     */
+    private $navigatorHTML;
+
+    /**
+     * Noop, pseudo backward compatability only
+     * @return DBHTMLText
+     */
+    public function BetterNavigator()
```
Comment (jonom) [inline src/Extension/BetterNavigatorExtension.php]: Please chop this out, I'll do a major release

File: src/Extension/BetterNavigatorExtension.php [hunk 3]
```diff
@@ -76,14 +95,65 @@ public function BetterNavigator() {
             ]);
 
             // Merge with page data, send to template and render
-            $bNData = new ArrayData($bNData);
-            $page = $this->owner->customise(['BetterNavigator' => $bNData]);
-            return $page->renderWith('BetterNavigator\\BetterNavigator');
```
Comment (jonom) [inline src/Extension/BetterNavigatorExtension.php]: I haven't tested this but it looks like the controller scope has been lost here. Do the various links that include `$AbsoluteLink` in the template still work? Aside from that, if devs opt to make use of the include hooks such as `<% include BetterNavigator\BetterNavigatorExtraDevTools %>` they would probably be expecting the template to be scoped to the current controller.

**Comment by jonom** [comment]
> Could potentially add an isAjax() just to be sure? It's not exactly unheard of to return a full body in an XHR, albeit rare.

I think if a full body response is being sent that it should be fine to inject the BetterNavigator output even if it's ajax. Devs can utilise the `shouldShowNavigator()` method you added if they need to turn it off.

> Technically the only API change is the type declarations

I'm sure you're right but I like to err on the side of caution with SemVer. If I'm adding brand new isolated features I'm happy to do a point release, but generally if I'm touching any existing functionality (apart from a bug fix) I prefer to do a major release. I feel like SilverStripe core pushes the edges of SemVer because the release cycle means major releases are only permitted every few years - but I think modules can and maybe should play it safer. So let's do a major release for this.

Thanks for your work!

File: src/Extension/BetterNavigatorExtension.php (hunk 3 above)
Comment (unclecheese) [reply src/Extension/BetterNavigatorExtension.php]: Yeah, tough one to explain here, but I think I've got it sorted in my latest commit. Basically, what was happening is that due to the prerendering in `onBeforeActionHandler`, values were being cached in SSViewer that were returning unexpected results when the final render happened. Specifically, `$Me` was always coming back as `ArrayData`, which was killing my project.

I had a quick look and wondered why the controller was even scoped at all, ran a few tests, which worked, because `<% base_tag %>` was providing the `$AbsoluteLink` fallback, and all was good. But I see now why you need the controller involved.

Interestingly, `$arrayData->customise($controller)->renderWith(..)` worked fine, but when the controller was the parent scope, the value of `$Me` was just getting clobbered.

In any event, I've refactored to remove prerendering. All it does now is pull in the requirements in the before handler. You lose the template level `<% require %>`, but the whole thrust of this PR is to get rid of the reliance on templates, so I don't see how that will be a problem. Custom includes should render the same.

File: src/Extension/BetterNavigatorExtension.php (hunk 2 above)
Comment (unclecheese) [reply src/Extension/BetterNavigatorExtension.php]: done

File: composer.json (hunk 1 above)
Comment (unclecheese) [reply composer.json]: Yeah, bonehead move here. That should be `^4.2`

File: src/Extension/BetterNavigatorExtension.php [hunk 4]
```diff
@@ -8,82 +8,157 @@
...
+    /**
+     * @var bool|null
+     */
+    private $shouldDisplay = null;
+
+    /**
+     * Load requirements in before final render. When the next extension point is called, it's too late.
+     * @return void
+     */
+    public function beforeCallActionHandler(): void
```
Comment (jonom) [inline src/Extension/BetterNavigatorExtension.php]: I might be misunderstanding something but return type declarations are a php7 feature and SS4.2 only requires php5.6+ right - I think it's only SS4.5 onwards that will require php7? So if we're targeting 4.2 (or any currently released version) these declarations would have to go?

File: src/Extension/BetterNavigatorExtension.php (hunk 3 above)
Comment (jonom) [reply src/Extension/BetterNavigatorExtension.php]: I wonder if `$Me` was getting clobbered because you've taken a slightly different approach. When I customised the controller previously I sandboxed the array data in a BetterNavigator variable `customise(['BetterNavigator' => $bNData])`. You've merged the array data straight in (`$this->owner->customise($navigator)`), perhaps overriding some properties of the controller?

Comment (unclecheese) [reply src/Extension/BetterNavigatorExtension.php]: Yeah, could be. I think the newest approach makes more sense, from my perspective.

File: src/Extension/BetterNavigatorExtension.php (hunk 4 above)
Comment (unclecheese) [reply src/Extension/BetterNavigatorExtension.php]: Nah, 4.2 Is EOL as of 30 June. All new APIs should use PHP7 return type declarations. Not existing APIs, though! I've bumped the version constraint to reflect that.

Comment (unclecheese) [reply src/Extension/BetterNavigatorExtension.php]: 4.2 went EOL since the issue of this PR, so I've bumped it out of the dependency constraint.

**Comment by unclecheese** [comment]
Hi, @jonom .. I've remove the PHP7 typing for you.

**Comment by unclecheese** [comment]
Lowered the framework constraint, too. No reason to make it any higher than 4.0 if we're not using PHP 7.

File: README.md [hunk 5]
```diff
@@ -69,10 +69,6 @@ All content, scripts and CSS are loaded via the BetterNavigator.ss template, so
 
 The BetterNavigator.ss template's scope is set to the page that is being viewed, so any methods available in your page controller will be available in the BetterNavigator.ss template. This should allow you to add custom links by page type and introduce complex logic if you want to.
 
-## Known issues
-
- * Probably won't work in IE8 or lower.
```
Comment (jonom) [inline README.md]: WHAT WHY DID YOU REMOVE THIS? 😂

**Comment by jonom** [comment]
Thanks for this @unclecheese merged and tagged 5.0.0
